### PR TITLE
Feature/tg 180/matching improvement

### PR DESCRIPTION
**Pull request by victor-paumier**

### CODE REVIEW

File: src/Controller/BuddyController.php [hunk 1]
```diff
@@ -17,10 +17,15 @@
 class BuddyController extends BaseController
 {
     /**
-     * @Route("", name="buddy_index", methods={"GET"})
+     * @Route("", name="buddy_index")
      */
     public function index(UserInterface $user, BuddyRepository $buddyRepo): Response
     {
+        /*$section = null;
```
Comment (JeremieSamson) [inline src/Controller/BuddyController.php]: why those comments ?

File: src/Controller/MatchingController.php [hunk 2]
```diff
@@ -34,35 +35,39 @@ public function step1(UserInterface $user, UserRepository $userRepo): Response
...
+     *
      * @ParamConverter("international", options={"id" : "internationalId"})
      */
     public function step2(UserInterface $user, UserRepository $userRepo, MatchingHelper $matchingHelper, User $international): Response
     {
-        $this->denyAccessUnlessGranted('match', $international);
+        $this->denyAccessUnlessGranted(UserVoter::MATCH, $international);
         $locals = $userRepo->getLocalStudentsNotMatched($this->getSection($user));
-        $matchingHelper->setMatchingScore($international, $locals);
+        $scores = $matchingHelper->matching($international, $locals);
```
Comment (JeremieSamson) [inline src/Controller/MatchingController.php]: var used only once ?

File: src/Controller/SectionController.php [hunk 3]
```diff
@@ -58,16 +65,26 @@ public function new(ObjectManager $om, Request $request)
...
-        if ($editForm->isSubmitted() && $editForm->isValid()) {
-            $om->persist($section);
-            $om->flush();
+        if ($editForm->isSubmitted()) {
+            if ($editForm->isValid()) {
+                $om->persist($section);
+                $om->flush();
+
+                $this->addFlash('success', 'Section saved');
+                if ($this->isGranted(User::ROLE_ADMIN)) {
```
Comment (JeremieSamson) [inline src/Controller/SectionController.php]: Ternary ?

File: src/Controller/UserController.php [hunk 4]
```diff
@@ -41,34 +42,39 @@ public function index(UserInterface $user, UserRepository $userRepo): Response
...
+            if ($editForm->isValid()) {
+                $om->persist($user);
+                $om->flush();
 
-            return $this->redirectToRoute('user_index');
+                $this->addFlash('success', $translator->trans('profile.updated'));
+
+                return $this->redirectToRoute('user_index');
+            }
+            $this->addFlash('error', 'Their is some errors in the profile');
```
Comment (JeremieSamson) [inline src/Controller/UserController.php]: Typo

File: src/Entity/Section.php [hunk 5]
```diff
@@ -531,13 +552,19 @@ public function getMotivations(): Collection
         return $this->motivations;
     }
 
+    /**
+     * @return Section
```
Comment (JeremieSamson) [inline src/Entity/Section.php]: it's void

File: src/Entity/Section.php [hunk 6]
```diff
@@ -531,13 +552,19 @@ public function getMotivations(): Collection
...
     {
         if (!$this->motivations->contains($motivation)) {
             $this->motivations[] = $motivation;
         }
     }
 
+    /**
+     * @return Section
+     */
     public function removeMotivation(Motivation $motivation): void
```
Comment (JeremieSamson) [inline src/Entity/Section.php]: typehint is wrong and no need to have phpdoc

File: src/Entity/Section.php [hunk 7]
```diff
@@ -550,13 +577,19 @@ public function getHobbies(): Collection
         return $this->hobbies;
     }
 
+    /**
+     * @return Section
```
Comment (JeremieSamson) [inline src/Entity/Section.php]: function is void no need to have phpdoc

File: src/Entity/Section.php [hunk 8]
```diff
@@ -550,13 +577,19 @@ public function getHobbies(): Collection
...
+     */
     public function addHobby(Hobby $hobby): void
     {
         if (!$this->hobbies->contains($hobby)) {
             $this->hobbies[] = $hobby;
         }
     }
 
+    /**
+     * @return Section
```
Comment (JeremieSamson) [inline src/Entity/Section.php]: same

File: src/Entity/User.php [hunk 9]
```diff
@@ -245,10 +239,6 @@ class User extends BaseUser
      * @var int
      *
      * @ORM\Column(type="integer")
-     * @Assert\LessThan(
```
Comment (JeremieSamson) [inline src/Entity/User.php]: Assert greater than instead maybe ?

File: src/Migrations/Version20181123153916.php [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+
+declare(strict_types=1);
+
+namespace DoctrineMigrations;
+
+use Doctrine\DBAL\Schema\Schema;
+use Doctrine\Migrations\AbstractMigration;
+
+/**
+ * Auto-generated Migration: Please modify to your needs!
```
Comment (JeremieSamson) [inline src/Migrations/Version20181123153916.php]: you can add all this in previous migration, better to only have only one migration per PR

File: src/Repository/BuddyRepository.php [hunk 11]
```diff
@@ -15,7 +15,7 @@ public function __construct(ManagerRegistry $registry)
         parent::__construct($registry, Buddy::class);
     }
 
-    public function getMyBuddiesWaitingValidation(User $user): array
+    public function getMyBuddiesWaitingValidation(User $user)
```
Comment (JeremieSamson) [inline src/Repository/BuddyRepository.php]: why did you remove typehint here ?

File: src/Repository/BuddyRepository.php [hunk 12]
```diff
@@ -44,17 +44,25 @@ public function getMyBuddiesValidated(User $user)
         return $qb->getQuery()->getResult();
     }
 
-    public function getAllBuddies(Section $section)
+    public function getAllBuddies(Section $section = null)
```
Comment (JeremieSamson) [inline src/Repository/BuddyRepository.php]: typehint ?

File: src/Repository/BuddyRepository.php [hunk 13]
```diff
@@ -76,15 +84,32 @@ public function getBuddiesByUser(User $user, $isLocal = false)
         return $qb->getQuery()->getResult();
     }
 
-    public function getWarnedBuddies()
+    public function getBuddiesToRemove()
```
Comment (JeremieSamson) [inline src/Repository/BuddyRepository.php]: typehint ?

File: src/Security/BuddyVoter.php [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+
+use App\Entity\Buddy;
+use App\Entity\User;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Component\Security\Core\Authorization\AccessDecisionManagerInterface;
+use Symfony\Component\Security\Core\Authorization\Voter\Voter;
+
+class BuddyVoter extends Voter
+{
+    const CONFIRM = 'confirm';
```
Comment (JeremieSamson) [inline src/Security/BuddyVoter.php]: public ?private ?

File: src/Services/MatchingHelper.php [hunk 15]
```diff
@@ -2,240 +2,197 @@
...
-    /**
-     * @param $languageMentor
-     * @param $languageBuddy
-     * @param $languageWantedMentor
-     * @param $languageWantedBuddy
-     *
-     * @return int
-     */
-    private function languageEqual($languageMentor, $languageBuddy, $languageWantedMentor, $languageWantedBuddy)
+    private function languageScore($languageMentee, $languageMentor, $languageWantedMentee, $languageWantedMentor): int
```
Comment (JeremieSamson) [inline src/Services/MatchingHelper.php]: typehint

File: templates/Hobby/index.html.twig [hunk 16]
```diff
@@ -57,4 +57,4 @@
 
 {% block javascripts %}
     <script src="{{ asset('build/hobby.js') }}"></script>
-{% endblock %}
+{% endblock %}
```
Comment (JeremieSamson) [inline templates/Hobby/index.html.twig]: line break

File: templates/Matching/Elements/userSummary.html.twig [hunk 17]
```diff
@@ -20,140 +20,138 @@
...
-            </p>
-        {% else %}
+        <p class="account-status {% if ageEqual %}f-green{% endif %}">
             {% if ageEqual %}
-                <p class="account-status">
-                    <img src="{{ asset('images/nok.svg') }}" alt="">
-                    {{ 'matching.age'|trans }}
-                    : {{ user.age ~ ' ' ~ "matching.label_age"|trans }}
-                </p>
+                <img src="{{ asset('images/ok.svg') }}" alt="">
```
Comment (JeremieSamson) [inline templates/Matching/Elements/userSummary.html.twig]: why empty alt ?

File: src/Controller/MatchingController.php (hunk 2 above)
Comment (victor-paumier) [reply src/Controller/MatchingController.php]: done

File: src/Controller/SectionController.php (hunk 3 above)
Comment (victor-paumier) [reply src/Controller/SectionController.php]: done

File: src/Controller/UserController.php (hunk 4 above)
Comment (victor-paumier) [reply src/Controller/UserController.php]: done

File: src/Entity/Section.php (hunk 5 above)
Comment (victor-paumier) [reply src/Entity/Section.php]: done

File: src/Entity/Section.php (hunk 6 above)
Comment (victor-paumier) [reply src/Entity/Section.php]: done

File: src/Entity/Section.php (hunk 7 above)
Comment (victor-paumier) [reply src/Entity/Section.php]: done

File: src/Entity/Section.php (hunk 8 above)
Comment (victor-paumier) [reply src/Entity/Section.php]: done

File: src/Migrations/Version20181123153916.php (hunk 10 above)
Comment (victor-paumier) [reply src/Migrations/Version20181123153916.php]: done

File: src/Repository/BuddyRepository.php (hunk 11 above)
Comment (victor-paumier) [reply src/Repository/BuddyRepository.php]: wrong typehint.
: mixed instead?

File: src/Security/BuddyVoter.php (hunk 14 above)
Comment (victor-paumier) [reply src/Security/BuddyVoter.php]: default is public

File: templates/Hobby/index.html.twig (hunk 16 above)
Comment (victor-paumier) [reply templates/Hobby/index.html.twig]: done

File: templates/Matching/Elements/userSummary.html.twig (hunk 17 above)
Comment (victor-paumier) [reply templates/Matching/Elements/userSummary.html.twig]: done

File: src/Security/BuddyVoter.php (hunk 14 above)
Comment (JeremieSamson) [reply src/Security/BuddyVoter.php]: Do you use it somewhere else than in your class ?
